### DISCUSSION TITLE
What is the reason for having to use a reverse proxy?

### DISCUSSION THREAD

**Discussion by thefunkygibbon** [dani-garcia/vaultwarden#3718] [Q&A]
Hi all, 
What is with the requirement for having to use a reverse proxy to basically have access to vaultwarden?  I'm not sure what webserver is part of the product, but most webservers have a https server as part of it.   
You can't seem to even simply create an account using the built in http only server.   (without context it) Just seems a rather odd design choice

**Comment by stefan0xC** [comment]
It's not required but recommended. See the [Enabling HTTPS wiki page](https://github.com/dani-garcia/vaultwarden/wiki/Enabling-HTTPS) for how to configure the Rocket web server. 

As to the reason why it's not recommended: I think a part of it is that it's often better to use a dedicated program instead of using rockets builtin https support (ie. you'll have better control about the ciphers and other parameters) and another part would be the websocket support that was missing in Rocket until recently.

**Comment by zacknewman** [comment]
If you don't benefit from having a reverse proxy server, then I'd recommend not using one. There are a few reasons I can think of that one would want to use a reverse proxy server:

- Load balancing.
- OCSP stapling.
- Control of ciphers and TLS versions not supported by Rocket.
- Access control based on application-level information (e.g., paths).
- IPv6 and IPv4 capable.
- Binding to TCP port 443 when Vaultwarden is run under a UID that is not 0 (i.e., "root").
- Handling multiple HTTP/HTTPS services on the same IP address and TCP port.
- Integrated support for generating and renewing x.509 certificates.

Before there were two additional reasons that are no longer relevant:

- Rocket used to warn that the TLS integration was not production ready.
- Rocket didn't use to support WebSockets.

Not to project _my_ needs/desires onto you, but I now no longer use a reverse proxy for my setup. In my case only my wife and I use Bitwarden + Vaultwarden, so the first bullet point is not applicable.

Most browsers don't even check stapled OCSP information; although Firefox does which is the browser I use, but I rotate my x.509 certificate every month so I don't care too much about that. This removes the second point.

The TLS information Rocket does allow one to [control](https://rocket.rs/master/guide/configuration/#tls) is enough for me (e.g., I only allow TLS 1.3 ciphers), so the third bullet point doesn't apply to me.

The password manager is on its own L2/L3 network, and only the IPv6 and IPv4 addresses belonging to my wife's and my devices have access—this includes the IPs associated with the WireGuard configuration which allows she and I to access it from an external network. This is enforced at both the router/firewall level as well as the OS the password manager runs on. I additionally disable the admin page, so the fourth bullet point is inapplicable.

The fifth bullet is sort of an illusion since even if you have a reverse proxy server that binds to separate IPv6 and IPv4 addresses, behind the scenes it will only forward to the IP Rocket listens to. Also depending on the OS you use and how it is configured, Rocket might actually handle both protocols. For example on many Linux-based OSes, listening to `::` will not only listen to all IPv6 addresses (in a particular `ip-netns(8)`), but also all IPv4 addresses. In my case I host it on OpenBSD 7.3-stable which does not do that—which I personally prefer since IPv6 and IPv4 are entirely different protocols, but I digress. Additionally, I prefer binding to specific IPs that way I can run multiple daemons on the same TCP/UDP port in the same `rdomain(4)`. This removes the fifth bullet point.

The sixth bullet point is not an issue since the Bitwarden client applications save the URI, so adding 8000 or whatever TCP port > 1023 you want to listen to is not an issue.

The seventh bullet point is not a problem for me since I avoid SNI when I can and use separate IPs for different HTTPS services which allows me to have more granular access control at the IP level.

OpenBSD comes with `acme-client(1)`, and I already use that to renew almost a dozen x.509 certificates. In the past I have used Certbot and acme.sh which are both quite easy to use. This removes the last bullet point.

Do those reasons apply to you? Only you know the answer to that.
